test/test_images/latencymako/sender_receiver.go

```diff
@@ -194,38 +126,112 @@ func main() {
 
-	sentEvents.Events = make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages)
-	acceptedEvents.Events = make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages)
-	failedEvents.Events = make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages)
-	receivedEvents.Events = make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages)
+	// Initialize records maps
+	executor.sentEvents = &pb.EventsRecord{
+		Type:   pb.EventsRecord_SENT,
+		Events: make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages),
+	}
+	executor.acceptedEvents = &pb.EventsRecord{
+		Type:   pb.EventsRecord_ACCEPTED,
+		Events: make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages),
+	}
+	executor.failedEvents = &pb.EventsRecord{
+		Type:   pb.EventsRecord_FAILED,
+		Events: make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages),
+	}
+	executor.receivedEvents = &pb.EventsRecord{
+		Type:   pb.EventsRecord_RECEIVED,
+		Events: make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages),
+	}
+
+	// --- Create Vegeta attacker
+
+	client := http.Client{Transport: requestInterceptor{
+		before: func(request *http.Request) {
+			id := request.Header.Get("Ce-Id")
+			executor.sentCh <- sentState{eventId: id, at: ptypes.TimestampNow()}
+		},
+		after: func(request *http.Request, response *http.Response, e error) {
+			id := request.Header.Get("Ce-Id")
+			t := ptypes.TimestampNow()
+			if e != nil || response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusMultipleChoices {
+				executor.failedCh <- failedState{eventId: id, at: t}
+			} else {
+				executor.acceptedCh <- acceptedState{eventId: id, at: t}
+			}
+		},
+	}}
+
+	executor.attacker = vegeta.NewAttacker(
+		vegeta.Client(&client),
+		vegeta.Workers(workers),
+		vegeta.MaxWorkers(workers),
+	)
+
+	return executor
+}
+
+type requestInterceptor struct {
+	before func(*http.Request)
+	after  func(*http.Request, *http.Response, error)
+}
+
+func (r requestInterceptor) RoundTrip(request *http.Request) (*http.Response, error) {
+	if r.before != nil {
+		r.before(request)
+	}
+	res, err := http.DefaultTransport.RoundTrip(request)
+	if r.after != nil {
+		r.after(request, res, err)
+	}
+	return res, err
+}
+
+func (ex *senderReceiverExecutor) Run(ctx context.Context) {
+	// --- Warmup phase
+
+	printf("--- BEGIN WARMUP ---")
+	if warmupSeconds > 0 {
+		if err := ex.warmup(warmupSeconds); err != nil {
+			fatalf("Failed to run warmup: %v", err)
+		}
+	} else {
+		printf("Warmup skipped")
+	}
+	printf("---- END WARMUP ----")
+
+	printf("--- BEGIN BENCHMARK ---")
 
 	// Start the events receiver
 	printf("Starting CloudEvents receiver")
-	startCloudEventsReceiver(processReceiveEvent)
 
-	printf("Starting latency processor")
+	waitForPortAvailable(defaultCEReceiverPort)
 
-	// Start the events processor
-	go processEvents()
+	cancelCeReceiver, err := ex.startCloudEventsReceiver(ex.processReceiveEvent)
```

Comment:
Instead of returning the cancel func, why not take a context? Then the caller controls the cancel scope (and can pass in the parent context if receiver-specific cancellation isn't needed):

```go
if err := ex.startCloudEventsReceiver(ctx, ex.processReceiveEvent); err != nil {
  fatalf(("Failed to start CloudEvents receiver: %v", err)
}
```